proof/invariant-abstract/RISCV64/ArchKernelInit_AI.thy

```diff
@@ -237,81 +275,197 @@ lemma kernel_mapping_slots_0x100[simp]:
-  apply (simp add: table_index_riscv_global_pt_ptr)
-  apply (simp add: bit_simps kernel_elf_base_def kernelELFBase_def kernelELFPAddrBase_def
-                   Kernel_Config.physBase_def)
+lemma is_page_aligned_physBase: (* FIXME MK: move to Kernel_Config_Lemmas *)
+  "is_aligned physBase pageBits"
+  by (simp add: Kernel_Config.physBase_def pageBits_def is_aligned_def)
+
+lemma is_page_aligned_kernelELFPAddrBase: (* FIXME MK: does not mention physBase, but could be moved up *)
+  "is_aligned kernelELFPAddrBase pageBits"
+  unfolding kernelELFPAddrBase_def
+  by (fastforce intro!: is_aligned_add is_page_aligned_physBase
+                simp: kernelELFPAddrBase_def pageBits_def is_aligned_def)
+
+(* FIXME MK: move *)
+lemma aligned_mask_le_mask_minus:
+  "\<lbrakk> is_aligned x m; m \<le> n; n < LENGTH('a)\<rbrakk> \<Longrightarrow> x && mask n \<le> mask n - mask m" for x :: "'a::len word"
+  by (metis and_mask_less' is_aligned_after_mask is_aligned_neg_mask_eq'
+           mask_2pm1 mask_sub neg_mask_mono_le word_less_sub_le)
+
+(* FIXME MK: rename, move *)
+lemma aligned_mask_plus_bounded:
+  "\<lbrakk> is_aligned x m; m < n; n < LENGTH('a) \<rbrakk> \<Longrightarrow> (x && mask n) + 2^m \<le> 2^n" for x :: "'a::len word"
+  by (metis add_mask_fold and_mask_less' is_aligned_add_step_le is_aligned_after_mask is_aligned_mask
+            leD leI order_less_imp_le t2n_mask_eq_if word_less_sub_1)
+
+lemma kernelELF_plus_page:
+  "((kernelELFPAddrBase && mask toplevel_bits) + 2^pageBits) \<le> 2^toplevel_bits"
+  by (fastforce intro: aligned_mask_plus_bounded is_page_aligned_kernelELFPAddrBase
+                simp: toplevel_bits_value bit_simps)
+
+lemma elf_window_4k:
+  "\<lbrakk> kernel_elf_base \<le> vref; vref < kernel_elf_base + (1 << pageBits) \<rbrakk> \<Longrightarrow>
+    table_index (pt_slot_offset max_pt_level riscv_global_pt_ptr vref) = elf_index"
+  using is_page_aligned_kernelELFPAddrBase
+  apply (simp add: table_index_riscv_global_pt_ptr elf_index_value toplevel_bits_value)
+  apply (unfold kernel_elf_base_def kernelELFBase_def bit_simps)
+  apply (simp add: add_ac)
+  apply (drule order_less_le_trans)
+   apply (rule word_plus_mono_right)
+    apply (rule kernelELF_plus_page[unfolded bit_simps toplevel_bits_value, simplified])
+   apply (simp add: pptrTop_def)
+  apply (simp add: bit_simps pptrTop_def mask_def is_aligned_nth)
   apply word_bitwise
-  apply (clarsimp simp: word_size)
+  apply clarsimp
   done
 
-lemma kernel_mapping_slots_0x1FE[simp]:
-  "0x1FE \<in> kernel_mapping_slots"
+lemma leq_elf_index:
+  "0x100 \<le> elf_index"
+  by (simp add: elf_index_value)
+
+lemma kernel_mapping_slots_elf_index[simp]:
+  "elf_index \<in> kernel_mapping_slots"
   by (simp add: kernel_mapping_slots_def pptr_base_def bit_simps pt_bits_left_def level_defs
-                pptrBase_def canonical_bit_def)
+                pptrBase_def canonical_bit_def leq_elf_index)
+
+lemma rewrite_eq_minus_to_plus_eq:
```

Comment:
this can be moved to Lib as well

Reply:
It definitely should be moved, but I'm struggling to find a good location. So far the best I've got is either sticking it in `Lib.Lib` or in `Word_Lib.Word_Lemmas_Internal`. An alternative might be `Word_Lib.More_Bit_Ring` from https://github.com/seL4/l4v/pull/587, once that is merged.

Reply:
It should not go into Word_Lib, it's not really much to do with words. I would actually be happy to keep it here, because it's just fixing a specific form so that a specific rewrite goes the way I need it to. The general case of such equations where the rest is solvable algebraically should already go away by simp (as does this). We could even make it a private lemma if we wanted to.